Request-a-copy improvements: Support access by secure link

kshepherd opened the pull request:
## References
* Requires [DSpace#10407](https://github.com/DSpace/DSpace/pull/10407)
* Related to REST contract: https://github.com/DSpace/RestContract/pull/307

## Description
Request a copy is an existing feature in DSpace, whereby users can request access to files they are not authorized to download, and the item author or a delegated helpdesk can approve or deny these requests.

Up until 8.x, all approved requests resulted in an email sent to the requester as attachments.

This is not a suitable solution for large files and doesn’t allow as much control over future authorization to those resources.

In this improvement, a secret key is generated to provide limited access to the requested bitstream(s) with a URL, rather than sending files as attachments. This link can be configured to allow access to just selected bitstreams, or all bitstreams of an item, and it can be configured with an “access period”, after which the secret URL will no longer work.

Example "access by secure link" item page:
![screenshot_2025-02-13_15:31:09_selection](https://github.com/user-attachments/assets/85ebc531-155d-4bb1-ab82-c63216d855f1)

## Notes for Reviewers
To test, you will also need the DSpace pull request applied: https://github.com/DSpace/DSpace/pull/10407

This is halfway between an improvement and a new feature. For those doing practical testing, I recommend reading the attached PDF. 
[Request-a-copy-secure-links.pdf](https://github.com/user-attachments/files/18785051/Request-a-copy-secure-links.pdf)

**There is a flyway sql migration** - two columns are added, so make sure you are testing on an instance that doesn't need to quickly roll back to an earlier version with the existing requestitem table.

**There is a new captcha added** - A new GDPR complicant, proof-of-work captcha is added to the "request a copy" form. I would like to propose we make this the default for Register User as well, and am happy to open a new PR accordingly. It is under the MIT license: https://altcha.org/captcha/

## Instructions for Reviewers
1. Read the supplied PDF: [Request-a-copy-secure-links.pdf](https://github.com/user-attachments/files/18785051/Request-a-copy-secure-links.pdf)
2. Check out or merge this PR into your local environment, edit configuration to suit the way you wish to test, and build + deploy
    a. If you are a non-developer tester, you may be able to follow the docker testing instructions instead
3. Make some items with a mix of restricted and public files, of various sizes (depending on the threshold you set, if any)
4. As anonymous and/or authenticated (but not authorized) users, make requests for the bitstreams or 'all files of items' in question. Take note of how the request form operates, ensuring that the captcha works well (it does not require human interaction).
5. As the owner or helpdesk delegate, check for the item request (if you are using a test environment and have mail server disabled in your config, you can see the email and copy the link from your log files). Follow the link to the approve/deny form
6. Try approvals and denials for requests as an owner or helpdesk delegate, with different access periods.
7. Check emails (or logs if disabled) for both kinds of emails, depending on your configuration. You may have some smaller files sent as attachments, larger ones sent as a secure link instead, you may have everything configured to send as a link, and so on. 
8. Test out access to the bitstreams / item page using the secure link provided. Note the behaviour of downloading a bitstream explicitly allowed by this link, compared to a bitstream the user already had access to, compared to a bitstream that is still unauthorized (and which should lead to a new request copy form)
9. Try accessing files with the link outside of the access period
10. Try images and videos with the media viewer enabled

List of changes in this PR:
1. New components added to support secure access (via secret URL) to item page with file section and bitstream download links
2. New components for media viewers (video and image) to support access via secret URLs
3. New Altcha GDPR compliant captcha for the request item form
4. Changes to the approve/deny form to allow access period to be determined
5. Routing changes to support item and itemrequest resolution in new components

## Attribution
Developed by The Library Code with the support of Technische Universität Berlin

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You do not need to complete this checklist prior creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [ ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`
- [x] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)
- [x] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [x] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

kshepherd commented:
Just a note to say that it's taken me a little while longer to fit in the google/altcha "generic" recaptcha for this, register user, feedback etc, than I first anticipated -- if it takes too long I will pull back on that "bonus extra" and we can consider this email form as just (optionally) protected by Altcha

MW3000 commented:
I tested the main functionality and it works great and as advertised ✅. I will test the different configuration possibilities in the next days.

MW3000 commented:
I did thoroughly test the different configuration options:

# Test basic configuration
- Request all files (of item) in restricted access, grant access => Works ✅
- Request one of many files, grant access => Works ✅
- Deny access request => Works ✅

# Test the captcha
=> Works ✅
(don't forget to set `altcha.hmac.key` when testing)

# Test using a size limit
Request single file below limit
=> Works ✅
Request single file above limit
=> Works ✅
If 'all files' was indicated in the request, this threshold will be activated if any of the files meet the configured size minimum.
=> Works ✅

# Test if access is closed again, after end of access period
=> Works insofar as access is no longer possible ✅
💡 But, after the end of the access period, I just get presented with a login page. It might be a better UX to show the custom download page and add a hint, that the access period expired?

# Test restricted images and videos with media viewer enabled
=> Image Viewer works on download page ✅
💡 Should be ported to work on regular item page too! (unrelated to this issue! but we could perhaps use the solution from this pull request?)

=> Video player works on download page ✅
💡 Should be ported to work on regular item page too! (unrelated to this issue! but we could perhaps use the solution from this pull request?)

kshepherd commented:
Thanks for your comments @tdonohue - I will work to address your specific points.

Regarding the duplication of components, you are correct; it's a tradeoff where we took the option of "maintain very similar / near-duplicates of some common components, and avoid adding all the additional if tests / access token handling to those core common components".
Some repositories that never enable request-a-copy might wonder why they now have to deal with / maintain / execute all these extra lines in components like item page, file section, download link and so on...
I'm happy to accept that my original approach is a little too conservative, and to go with the consensus. I'd prefer that we still kept a separate 'item page' so it can be modified and presented in a different way to a "normal" item page, but we can probably consolidate the code in:
* full file section
* download link
* media viewers

Does that sound reasonable?

tdonohue commented:
@kshepherd : I see. I'm OK with the Item pages being different, provided we find a way to document (in code comments of the component) why we have a separate copy of the Item page component.  

I agree though that I'd rather see the other individual components (that you've listed) consolidated to avoid code duplication.  That way if we ever update the media viewers (or similar), we aren't having to do so in multiple places.

I do also want to point out that Request a Copy seems to be **enabled by default** as of DSpace 7.  So, I expect *most/many* repositories do use it now (unless they've specifically turned it off).

kshepherd commented on src/app/app-routing-paths.ts:
```diff
@@ -34,6 +34,48 @@ export function getBitstreamRequestACopyRoute(item, bitstream): { routerLink: st
+ * @param item_uuid item UUID
+ * @param accessToken access token (generated by backend)
+ */
+export function getAccessTokenRequestRoute(item_uuid, accessToken): { routerLink: string, queryParams: any } {
```
getAccessTokenRequestFileRoute was unused, I've removed it thanks

kshepherd commented on src/app/bitstream-page/bitstream-download-page/bitstream-download-page.component.ts:
```diff
@@ -83,45 +84,80 @@ export class BitstreamDownloadPageComponent implements OnInit {
     this.bitstream$.pipe(
       switchMap((bitstream: Bitstream) => {
         const isAuthorized$ = this.authorizationService.isAuthorized(FeatureID.CanDownload, isNotEmpty(bitstream) ? bitstream.self : undefined);
+        // TODO isAuthorizedByToken check here so we already know if this token is going to be valid?
```
a todo that was already resolved in code yes, thanks - removed

kshepherd commented on src/app/bitstream-page/bitstream-download-page/bitstream-download-page.component.ts:
```diff
@@ -83,45 +84,80 @@ export class BitstreamDownloadPageComponent implements OnInit {
       map((data) => data.bitstream));
 
     this.bitstream$ = this.bitstreamRD$.pipe(
+      // TODO: this redirect was commented out earlier...
```
a note to myself, removed! sorry - this component obviously slipped through my QA/cleanup process

ybnd commented:
@kshepherd @tdonohue 

From our point of view the duplicated Bitstream sections/links and Item pages are also a red flag, especially when taking into account theming and custom Entities.

> Regarding the duplication of components, you are correct; it's a tradeoff where we took the option of "maintain very similar / near-duplicates of some common components, and avoid adding all the additional if tests / access token handling to those core common components".  
> Some repositories that never enable request-a-copy might wonder why they now have to deal with / maintain / execute all these extra lines in components like item page, file section, download link and so on...

While I agree that limiting the duplication to Item components only is an improvement, I'd still argue that those components are some of the most commonly customized in DSpace as a whole.
IMO this trade-off leans heavily to reducing duplicate code, as the file section / download link code is a lot less likely to be modified in the first place.

I'd like to propose two alternatives to keep Item pages deduplicated:
1. Move the `accessToken` logic to a resolver and (optionally) use route data within the Item page components
    - Using a resolver would make it very easy to use the ItemRequest object in existing components without having to change their inputs at all
        - e.g. the warning could be moved to `ds-item-alerts` along with conceptually similar things
    - If the access token logic is limited to the components where it matters (warning, download links, ...), themed Item pages would just work with this feature out of the box
2. Rethink the `/access-by-token` page as completely separate from the Item page, and don't try to keep it "in sync" 
    - It could contain the file section and warning, along with links to the simple or full Item pages.
    - A clearly distinct title/layout could serve as another nudge for users _not_ to share the access link if they happen to gloss over the warning
    - If someone _wants_ to mirror the simple Item page here, they can always use theming

kshepherd commented on src/app/item-page/bitstreams/request-a-copy/altcha-captcha.component.ts:
```diff
@@ -0,0 +1,50 @@
+  schemas: [CUSTOM_ELEMENTS_SCHEMA],
+  standalone: true,
+})
+export class AltchaCaptchaComponent implements OnInit {
```
done

kshepherd commented on src/app/request-copy/grant-request-copy/grant-request-copy.component.html:
```diff
@@ -1,9 +1,36 @@
+<!--      </a>-->
+        <a [attr.routerLink]="previewLinkOptions.routerLink" class="dont-break-out d-block" [target]="'_blank'"
+           [attr.queryParams]="previewLinkOptions.queryParams"
+           [attr.rel]=""
```
done

kshepherd commented on src/app/request-copy/grant-request-copy/grant-request-copy.component.html:
```diff
@@ -1,9 +1,36 @@
+      <p>{{ 'grant-request-copy.intro.link.preview' | translate }}
+<!--      <a [routerLink]="previewLinkOptions.routerLink" class="dont-break-out d-block" [queryParams]="previewLinkOptions.queryParams" [target]="'_blank'">-->
+<!--        {{ previewLink }}-->
+<!--      </a>-->
```
done

kshepherd commented on src/assets/i18n/en.json5:
```diff
@@ -2070,10 +2078,26 @@
+
+  "grant-request-copy.access-period.86400": "1 day",
+
+  "grant-request-copy.access-period.604800": "7 days",
```
I see what you mean... this seemed like a pretty easy "set and forget" way to do it without trying to put too much extra work in the code, but I'm happy to do something there, and then have keys like:

en:
`grant-request-copy.access-period-unit.day`: "Day"
`grant-request-copy.access-period-unit.days`: "Days"
`grant-request-copy.access-period-unit.week`: "Week"
`grant-request-copy.access-period-unit.weeks`: "Weeks"
`grant-request-copy.access-period-unit.month`: "Month"
`grant-request-copy.access-period-unit.months`: "Months"
`grant-request-copy.access-period-unit.year`: "Year"
`grant-request-copy.access-period-unit.years`: "Years"

(or year -> year_plural, whatever looks better to translators)

de:
`grant-request-copy.access-period-unit.day`: "Tag"
`grant-request-copy.access-period-unit.days`: "Tage"
`grant-request-copy.access-period-unit.week`: "Woche"
`grant-request-copy.access-period-unit.weeks`: "Wochen"
`grant-request-copy.access-period-unit.month`: "Monat"
`grant-request-copy.access-period-unit.months`: "Monate"
`grant-request-copy.access-period-unit.year`: "Jahr"
`grant-request-copy.access-period-unit.years`: "Jahre"

It would just mean I can't say, e.g. "a fortnight" or "a semester",  if that is what my admin wanted, or force the system to say "1.5 Years" instead of "18 Months" or "3 and a half months" instead of... i guess if exact months (or any unit) can't be calculated exactly, you would drop to the lower unit and try that instead? 
I thought my initial approach just gave a bit more flexibility - the label can be anything, the admin just has to decide how many seconds that is.

Regarding the sidenote, yes, it's a little silly and theoretical and not configured by default. But is nice for testing and helping to show the relation between the numbers to time units. Happy to remove it.

kshepherd commented:
thanks @ybnd , I see the points and I'll think about the suggestions.. a resolver might be a good approach. Do you mind if I ping you for specific feedback on that if I try something out?

regarding suggestion #2 - that is kind of what i am going for in the current code, but i think i see your certain, that any breaking changes to ItemComponent would also affect these pages?

kshepherd commented:
@tdonohue @ybnd just another note to say i do like a resolver as a way of separating and simplifying the way of providing the ItemRequest object to whatever component is ultimately loaded, but in addition:

* all entity types are supported.. the custom "access by token item page" extends item *component* and acts a bit liek a cross between a simple and full item page -- because i want the full file section and don't want to use the listableObjectComponentLoader to render an item in viewmode (i don't want to modify that class to expect access tokens, it is out of scope except for this situation) -- but the item page routes run for all entity types and untyped item anyway, so that is not an issue

* if i could use the ThemedItemPageComponent with an if test to either include the full or 'view mode' file section, it'd be nice, but unfortunately it hands everything off to the generic listable object component loader.
If i do any extra template work here, i'd have to test for access token and then include either a whole custom item view, or the listable object viewer -- or, i direct to ThemedFullItemPageComponent instead (which does have nicer separation of sub-components in an item viewer) and just do some templating to make the metadata look a bit more like the simple item view... 

So I'm a bit stuck of the smoothest way forward on this... yes, I want to farm out all the parts of the item 'view' except for the file section... perhaps it's enough to hide the file section in an item view mode where an access token is present, and include the full file section in ThemedItemPageComponent afterwards, where it is present, but that means modifying this super-generic "ListableObjectComponentLoader" class to handle accessTokens when we only care about it in this super-specific "the object is an item and we are calling it in Standalone View Mode" case...

kshepherd commented:
@tdonohue @ybnd I have pushed a refactored version (I still have the original branch here to compare if/when we need it) -- it has a little console output still included so please ignore that, but I submit it to get your approval on the new approach:

1. A resolver will resolve any accessToken query param to an ItemRequest object, in the parent item page route path
2. A new `ItemWithSupplementaryData` (name TBD, just there as an example) extends Item with an optional `ItemRequest` var
3. The `MediaViewerItem` component was changed to also contain an optional `accessToken` string, for building URLs
4. The ItemPageComponent, if `itemRequest` is present from the resolver, will replace the `itemRD.payload` with a new one using the new object model above - this means `Item` will still work as normal, but if it happens to be `instanceof ItemWithSupplementaryData`, we can use the itemrequest data
5. The `FileSection` component, depending on whether item request data exists, will use either the normal or secure download link component to render the download link (_note: i am happy to consolidate this into the main download link too, but this does have quite a bit of extra testing just to determine paths, icons, etc so for now i kept it separate_)
6. The Image and Video media viewers will look for an accessToken in its `MediaViewerItem` data and use it to construct the URLs, if present

Is this closer to an acceptable solution?

tdonohue commented:
NOTE: This PR has merge conflicts caused by the merger of #3997.  You will need to rebase this PR on the latest `main` and update it to use ["control flow" syntax](https://angular.dev/guide/templates/control-flow) by replacing any `ngIf`, `ngFor`, or `ngSwitch` usages in HTML templates with `@if`, `@for` or `@switch`.  See the control flow docs or the above PR for more details.

kshepherd commented on src/app/item-page/access-by-token/item-access-by-token-view.component.ts:
```diff
@@ -0,0 +1,128 @@
+    //MetadataRepresentationListComponent,
+  ],
+})
+export class ItemAccessByTokenViewComponent extends ItemComponent implements OnInit {
```
as per recent comment, this view is no longer used, and the standard item page component (extended by all entity types), file section and media viewer is now used

ybnd commented on src/app/item-page/simple/field-components/file-section/file-section.component.html:
```diff
@@ -14,7 +15,21 @@
+              @if (!last) {
+                <span innerHTML="{{separator}}"></span>
+              }
+            </ds-file-download-link>
```
The only thing that's different here is the component selector, and it's not likely we'd need to make any other changes at this level. I'd advise to move this "choice" into the link component.

kshepherd replied:
I've done this and hopfully restored the file section to its former self

ybnd reviewed (changes requested):
The resolver works perfectly, but if take things a step further with how you handle the data it returns everything could be even simpler & leaner IMO

kshepherd commented on src/app/item-page/access-by-token/field-components/file-download-link/item-secure-file-download-link.component.ts:
```diff
@@ -0,0 +1,153 @@
+   * @param itemRequest the item request object
+   */
+  getAccessByTokenBitstreamPath(itemRequest: ItemRequest) {
+    return getBitstreamDownloadWithAccessTokenRoute(this.bitstream, itemRequest.accessToken);
```
Yes thanks, this better articulates the approach I first intended to take. I used (activated) route snapshot as it appeared safe to use by the time the bitstream component is loaded (follows similar patterns used in eg edit bitstream) and avoids unnecessary async operations

kshepherd commented on src/app/core/shared/item-with-supplementary-data.model.ts:
```diff
@@ -0,0 +1,20 @@
+ * to help components determine how the Item or its data/bitstream should be delivered
+ * and presented to the users, but not part of the actual database model.
+ */
+export class ItemWithSupplementaryData extends Item {
```
Thanks, I've done this to hopefully simplify things further.

ybnd replied:
This class is still here; as far as I can tell it's only used in one other place as part of an `instanceof` check that isn't needed anymore?

kshepherd commented:
@tdonohue @ybnd @paulo-graca thanks all for your feedback and reviews so far. This is ready for a another review as I believe I have addressed all feedback. I have significantly rewritten and simplified a lot of areas:

* Access periods are now delta strings like +1DAY, +4WEEKS, FOREVER, to match existing patterns and are able to be parsed by the same backend services. These are used in the form, but the actual access expiry will be parsed and stored as a SQL TIMESTAMP in the backend.
* All the duplicated components are removed in favour of a simple routing-based approach, with some small changes to item and file download link components, media viewer components, to include access tokens when downloading if present (and small changes to fontawesome icons displayed beside links if an access token is present and valid for a restricted file)
* Other feedback addressed (inline doc, code style, code cleanup, etc)

tdonohue commented on src/app/item-page/simple/item-page.component.html:
```diff
@@ -3,6 +3,16 @@
+          <ng-container *ngVar="(itemRequest$ | async) as itemRequest">
+            @if (hasValue(itemRequest)) {
+              <div class="alert alert-warning wb-100 mb-2">
+                <p><span role="img" class="request-a-copy-access-icon"><i class="fa-solid fa-lock-open"></i></span>{{'bitstream-request-a-copy.access-by-token.warning' | translate}}</p>
```
The open lock icon needs to have `alt` text added.  Currently, if you scan this page with the "axe DevTools" plugin for Chrome, it gives this error:
```
Ensure [role="img"] elements have alternative text
```

kshepherd replied:
Done for this and bitstream link, using `[attr.aria-label]` so a i18n key can be used.

tdonohue reviewed (changes requested):
@kshepherd : Gave this another review & test today.  The main results of the test were reported in my [backend review](https://github.com/DSpace/DSpace/pull/10407#pullrequestreview-2715041922). 

I did notice a few minor things that could use cleanup in this frontend PR (see inline comments below).  Most of them are extremely minor.  The most important thing to fix is to ensure the accessibility issue (with the unlock icon) is resolved.  Beyond that, this frontend PR looks good to me.

kshepherd commented:
@tdonohue thanks! I have addressed all these points.

tdonohue reviewed (approved):
👍 Thanks @kshepherd !  This frontend PR now looks good to me.  As noted on the backend though, there's still a tiny bug related to `+1WEEK`.  

Finally, please note this has a new (minor) merge conflict to fix up.

ybnd commented on src/app/core/auth/access-token.resolver.ts:
```diff
@@ -0,0 +1,57 @@
+  return itemRequestDataService.getSanitizedRequestByAccessToken(accessToken).pipe(
+    getFirstCompletedRemoteData(),
+    // Handle authorization errors, not found errors and forbidden errors as normal
+    redirectOn4xx(router, authService),
```
Right now this means that you'll be shown a login page if you try to follow the access link after it expires.

It would be more informative if we could still show the Item page, but with a notice to explain what's going on.

ybnd commented on src/app/item-page/simple/item-page.component.html:
```diff
@@ -3,6 +3,17 @@
+                }
+              </div>
+            }
+          </ng-container>
```
Could this be moved to a separate component like the other alerts below?

If we change the REST/resolver logic to still expose expired `RequestItem`s this would also be the place to show an alternative message, e.g. something along the lines of "you're viewing this page with an expired access token, so you no longer have access to certain files"

The date is shown as an ISO timestamp, which is a bit hard to read. Could you transform it with e.g. `| date:"YYYY/MM/dd hh:mm"`?
![20250327-141723](https://github.com/user-attachments/assets/e0555918-3122-4f76-87ef-a4a7811715b3)

kshepherd replied:
done

ybnd commented on src/app/item-page/simple/item-page.component.scss:
```diff
@@ -4,3 +4,7 @@
 }
+.request-a-copy-access-icon {
+  margin-right: 4px;
+  color: #26a269;
```
This colour is defined in two places.

Could you use a Bootstrap variable like `--bs-success` for it, or create a new `--ds-` variable in `_custom_variables.scss` to make sure it's easy to theme?

ybnd commented on src/app/request-copy/email-request-copy/email-request-copy.component.html:
```diff
@@ -12,18 +13,46 @@
+        <!-- Access period dropdown -->
+        <div ngbDropdownMenu aria-labelledby="accessPeriod">
+          @for (accessPeriod of (validAccessPeriods$ | async); track accessPeriod) {
+            <button
```
This dropdown looks very weird in the `custom` theme:
![20250327-141128](https://github.com/user-attachments/assets/1977d9fd-19c0-4a1a-a7ee-e9e21e4dde25)

...vs. in the `dspace` theme:
![20250327-141325](https://github.com/user-attachments/assets/d6a4838e-c0a2-4855-844b-7776def3e509)

kshepherd replied:
i hadn't seen this and it just followed the same pattern of ngbDropdownMenu used elsewhere but I'm trying a few things out, please let me know if you still see that

ybnd reviewed (changes requested):
@kshepherd @tdonohue sorry, I found a few more things to bring up

kshepherd commented:
@ybnd just pushed changes now.

- New accessExpired flag can be checked along with acceptRequest, so expired or not-granted links  will show an error and draw 'normal' download links instead of doing a hard 4xx redirect
- The download links will just be the usual restricted ones, so if a token has expired, the user can just click to submit a new request-a-copy request
- Old ItemWithSupplementaryData component removed
- Notifications moved to sub-component of item page and display a warning on valid access token, or a danger alert if expired or not-granted, with an appropriate error message
- Tests written for each scenario to ensure correct alert is shown
- Date formatted in messages as yyyy-MM-dd

You can see examples of the "expired" and "not granted / revoked" errors below:

![screenshot_2025-03-27_16:44:20_selection](https://github.com/user-attachments/assets/0c28e7d9-2cfa-4685-a6ff-44fc14573362)
![screenshot_2025-03-27_16:37:11_selection](https://github.com/user-attachments/assets/a6658427-5475-4154-9d09-161438adee68)

ybnd reviewed (changes requested):
Many thanks for the quick turnaround, and sorry again for the short notice.

Very happy with the overall shape of the codebase & UX now 👍

The dropdown issue in the `custom` theme needs one more tiny change to fully work and then we can merge!

pnbecker commented on src/themes/custom/app/request-copy/email-request-copy/email-request-copy.component.ts:
```diff
@@ -13,7 +16,7 @@ import { BtnDisabledDirective } from '../../../../../app/shared/btn-disabled.dir
   templateUrl: './../../../../../app/request-copy/email-request-copy/email-request-copy.component.html',
   standalone: true,
-  imports: [FormsModule, NgClass, TranslateModule, BtnDisabledDirective],
+  imports: [FormsModule, NgClass, TranslateModule, BtnDisabledDirective, AsyncPipe],
```
Thank you @ybnd very much. I cherry picked your commit. I currently cannot test this, but I expect you tested your change?

kshepherd commented:
thanks @ybnd , as you see the change is cherry-picked, and I tested it in my custom theme and it's looking fine
![screenshot_2025-03-28_13:43:39_selection](https://github.com/user-attachments/assets/dbba4d75-db5f-4923-aa11-d0be31886a71)

tdonohue commented:
Thanks all!  As I see that @ybnd gave this a 👍 in his [last comment](https://github.com/DSpace/dspace-angular/pull/3984#pullrequestreview-2724934411), and his one minor suggestion has been accepted, I'm considering this at +2.  I see the backend PR is also at +2.

Therefore, I'm merging this immediately.  Thanks again @kshepherd and @pnbecker for all the hard work on this!

Keep in mind, we will need documentation both for the Request a Copy enhancements and the Altcha captcha options added to https://wiki.lyrasis.org/display/DSDOC9x/  Therefore, I'm adding the `needs documentation` label until that is complete.

tdonohue commented:
@kshepherd or @pnbecker : Would one of you be able to add documentation for this feature in our 9.0 docs? https://wiki.lyrasis.org/display/DSDOC9x   Ideally, basic documentation would exist prior to our [9.0 Testathon](https://wiki.lyrasis.org/display/DSPACE/DSpace+Release+9.0+Testathon+Page) (which starts on Monday, April 7).  But, if it cannot exist by then, it should hopefully be created shortly thereafter.

tdonohue commented:
Docs have now been added to https://wiki.lyrasis.org/display/DSDOC9x/Request+a+Copy and https://wiki.lyrasis.org/display/DSDOC9x/CAPTCHA+Verification
